followthemoney/mapping/entity.py

```diff
@@ -112,16 +115,24 @@ def map(
         # from that accessible to phone and address parsers.
         for prop in self.properties:
             if prop.prop.type == registry.country:
-                prop.map(proxy, record, entities)
+                discarded_values = prop.map(proxy, record, entities)
+                for value in discarded_values:
+                    log.warn(f"[{self.name}] Discarded unclean value \"{value}\" for property \"{prop.prop.qname}\".")
```

Comment:
@pudo You suggested also logging the proxy ID here. I can see how this can be useful, especially when the mapping uses `id_column`. I just don’t know how to implement this in a smart way as the proxy ID hasn’t been computed yet at this point (and it cannot be computed earlier as that requires knowing the property values).

Only option I see is not to log the warning message directly, but cache them until the ID has been computed, something like this:

```python
discarded_values: dict[Property, set[value]] = defaultdict(set)

for prop in self.properties:
  if prop.prop.type == registry:country:
    # Save discarded values
    discarded_values = prop.map(proxy, record, entities)
    discarded_values[prop].update(discarded_values)

for prop in self.properties:
  if prop.prop.type == registry:country:
    # Save discarded values
    discarded_values = prop.map(proxy, record, entities)
    discarded_values[prop].update(discarded_values)

proxy.id = self.compute_key(record)
if proxy.id is None:
  if self.id_column:
    log.warn(...)
  if self.keys:
    log.warn(...)
  return None

if discarded_values:
  # Finally about discarded values here
  for prop, values in discarded_values.items():
    log.warn(...)
```

But that’s a lot of code and indirection for a simple log message… Any good ideas?

Reply:
Yeah I think it's too much faff in the core loop of the mapper. Let's leave it.